update seek to seek_prefix

blasrodri opened the pull request.

mina86 committed:
store: introduce temperature and NodeStorage abstraction (#7332)

Add Temperature enum and NodeStorage struct which introduces a layer
between node’s storage and Store interface.  As it stand, this is
really just complicating code, however, the eventual goal is to have
hot and cold storage with NodeStorage being interface to access
them.  Store struct will offer access to either of those databases.

The Temperature enum is a ZST so in the current form it should be
optimised out at compile time.

matklad committed:
refactor: remove unused no-op runtime "cache" (#7508)

Contract cache now lives in contract-runtime, this code is just an
elaborate indirection, we can get rid of it.

mina86 committed:
store: add ColdDatabase Database interface (#7482)

Introduce ColdDatabase struct which wraps a real Database interface
and transparently translates reads and writes as expected by the users
of Database interface to reads and writes as they will happen in cold
storage.  This is needed because we want to optimise cold storage
slightly (e.g. by simplifying key in State column and removing
reference counts).

jakmeier committed:
fix: rocksdb flush all column families (#7484)

Standard rocksdb `flush()` only flushes the default column family.
See https://github.com/facebook/rocksdb/blob/95ef007adc9365fbefc0f957722a191c1fd7dcd3/include/rocksdb/db.h#L1398-L1400

To flush all column families as intended, iterator over them and flush
them individually.

This change only affects the runtime parameter estimator. In the
productive client, we never force a flush. (As of today.)

mina86 committed:
nightly: switch to fuzzing 1.28.1 and 1.29.0 (#7514)

Firstly, 1.29 release cycle has started so start fuzzing 1.29.0.

Secondly, 1.28.1 has been released so switch to fuzzing that rather
than 1.28.0.

And lastly, with all that don’t waste time fuzzing 1.27 any longer.

matklad committed:
refactor: cleanup contract runtime entry point (#7510)

Just minor cleanups:

* remove needles `&mut`
* flatten the entry point a bit
* collapse two unwraps into one

mina86 committed:
store: refactor flat_state.rs to reduce amount of annotations (#7513)

Add flat_state::maybe_new function which creates FlatState object when
requested and the feature is enabled.  This removes bunch of
annotations and noise from shard_tries.rs.

Furthermore, split definition of FlateState to one when the feature is
enabled and another when it’s disabled.  This further removes bunch of
annotations this time at definition place.

Furthermore, by making the latter impossible to instantiate (by using
an enum with no variants) this encodes in the type system that the
flat state won’t be created if the feature is not enabled and its
method won’t be called.  This also makes `Option<FlatState>` a ZST.

mina86 committed:
chain: pass the raw Database when creating PeerManagerActor (#7459)

Since PeerManager has uses the database in a way noticeably different
than the rest of the code (that is, it just stores mutable data there
rather than appending new information to the growing chain), PeerStore
bypasses the Store abstraction.

Despite that, when creating a new PeerManagerActor a Store instance
needs to be passed as constructor argument.  The Store abstraction is
then pealed away to get the inner Database object.

Change the code so that the Database is passed directly and pealing of
the abstraction happens at the call site.

This will also simplify future code which adds NodeStorage and
Temperature abstractions.

mzhangmzz committed:
Improve the logic of handling old protocol versions (#7501)

This PR introduces two changes:
1. Complete the logic that make nodes panic if their protocol version is old. Previously, the check was only triggered at epoch boundaries. If validators restart their nodes, their nodes will keep running. 
2. Introduce a new protocol feature that forbids blocks produced with old protocol versions. This can help kick out validators that are not upgraded.

mzhangmzz committed:
Simplify simple_nightshade shard layout config (#7499)

This PR simplifies the code around simple nightsahde shard layout upgrade. Instead of adding the hacky field `simple_nightsahde_shard_layout` in genesis, now the logic is moved to inside AllEpochConfig

mina86 committed:
network: introduce near_peer_message_sent_by_type_{bytes,total} metrics (#7523)

* network: introduce near_peer_message_sent_by_type_{bytes,total} metrics

* fmt

mina86 committed:
store: avoid Vec conversion when getting data from RocksDB (#7494)

Currently, when data is read from RocksDB it’s converted from
RocksDB’s internal type (specifically DBPinnableSlice) to a vector.

The problem is that we often end up discarding this vector just as
soon as it’s created.  For example, when deserialising objects using
get_ser method an owned vector is unnecessary.  And when accessing
DBCol::State the data is converted into an Arc which forces another
allocation and memory copy.

Introduce DBBytes type which is what Database::get_raw_bytes and
Database::get_with_rc_stripped methods return.  For RocksDB the object
holds RocksDB specific DBPinnableSlice which than caller can treat as
a slice or convert to a vector or an arc as desired.

Furthermore, change Store::get to return data in this form.
Furthermore, for convenience, introduce Store::get_vec which does
conversion to a vector.

pompon0 committed:
Made PeerActor route messages without contacting PeerManager (#7521)

* direct routed message

* fixed tests

* compilation fix

* applied comments

* removing unnecessary clones

* compilation fix

Co-authored-by: near-bulldozer[bot] <73298989+near-bulldozer[bot]@users.noreply.github.com>

mina86 committed:
pytest: fix one_val.py test (#7517)

The test broke after recent changes to tracked shards configuration
options.  Fix it.

robin-near committed:
Do not forward the same chunk part twice. (#7512)

Ekleog committed:
Fix compilation determinism issue (#7443)

There was a compilation determinism issue in our instrumentation, due to iterating over a HashMap when generating the thunks. With this, I add:
- A fuzzer that tries to automatically detect non-determinism instances
- An output from this fuzzer that generates said non-determinism; auto-generated so that when running `cargo test` the regression would be caught
- The fix for the compilation determinism issue per se

I've been running the fuzzer locally for a bit to check there's no other obvious instance of compilation non-determinism, and coverage has already doubled since the fuzzer found the crash I checked in so hopefully other instances of compilation non-determinism would be hard to find.

cc @akhi3030 @matklad

pompon0 committed:
adjusted log message (#7531)

because the test relies on the string in a log.
Passing nayduck run: https://nayduck.near.org/#/test/358900

nikurt committed:
Shardnet release schedule (#7529)

This change will need to be reverted before mainnet protocol version reaches 100.

mzhangmzz committed:
Upgrade shardnet shard layout (#7522)

Tested with localnet

jakmeier committed:
feat(estimator): remove IO costs in icount/qemu (#7527)

Set the cost of IO read and write bytes in qemu based estimations to 0.

The only real purpose of having qemu based measurements today
is having a stable output that can be used to detect performance
regressions. Since #7484 this no longer works because read IO proofs
to be unstable.

Setting IO costs to 0 should stabilize gas costs at least partially.

Also setting default estimation to `time` to better reflect how we use it now.

mina86 committed:
core: avoid double Vec::last call in TrieItorator::iter_step (#7533)

This avoids an `expect` present in the second invocation.

akhi3030 committed:
refactor: split out some more tests from `process_blocks.rs` (#7536)

See #7287 for similar work.

akhi3030 committed:
refactor: split out some more tests from `process_blocks.rs` (#7537)

See https://github.com/near/nearcore/pull/7287 for similar work.

This should be the final PR that is based splitting out tests based on `ProtocolFeature`.  I suspect that we could split some more tests from `process_blocks.rs` as well.  But we will have to figure out how to organise them.

mina86 commented on runtime/runtime/src/state_viewer/mod.rs:
```diff
@@ -144,7 +144,7 @@ impl TrieViewer {
+        iter.seek_prefix(&query)?;
         for item in &mut iter {
             let (key, value) = item?;
             if !key.starts_with(query.as_ref()) {
```
This check isn’t necessary any more.

mina86 commented on core/store/src/trie/update.rs:
```diff
@@ -224,7 +224,7 @@ impl<'a> TrieUpdateIterator<'a> {
             None => None,
         };
-        trie_iter.seek(&start_offset)?;
+        trie_iter.seek_prefix(&start_offset)?;
```
The `!key.starts_with(prefix)` check in `stop_cond` on line 264 is no longer necessary.

mina86 commented on core/store/src/trie/mod.rs:
```diff
@@ -948,7 +948,7 @@ mod tests {
         let trie = tries.get_trie_for_shard(ShardUId::single_shard(), root.clone());
         let mut iter = trie.iter().unwrap();
-        iter.seek(&vec![0, 116, 101, 115, 116, 44]).unwrap();
+        iter.seek_prefix(&vec![0, 116, 101, 115, 116, 44]).unwrap();
```
The test should now pass if we change `pairs[..2]` to just `pairs` on line 957.

blasrodri replied:
It fails

blasrodri replied:
will need to create a new element

blasrodri replied:
added on 591a2e4958a1e64993247ee2a66e4aa36d9403d0

mina86 replied:
No, that’s wrong.  It should pass without that element.  It doesn’t because the seek_prefix implementation isn’t correct.  Similarly, `assert!(key.starts_with(&query),` mentioned in another comment fails the test because the current implementation is incorrect.

mina86 commented on core/store/src/trie/iterator.rs:
```diff
@@ -451,7 +459,7 @@ mod tests {
         let mut iterator = trie.iter().unwrap();
-        iterator.seek(&seek_key).unwrap();
+        iterator.seek_prefix(&seek_key).unwrap();
         let result1: Vec<_> = iterator.map(Result::unwrap).take(5).collect();
```
How about:
```suggestion
        let iterator = iterator.map(Result::unwrap).inspect(|(key, _)| {
            assert!(key.starts_with(seek_key),
                    "‘{key:?x}’ does not start with ‘{start_key:?x}’");
        });
        let result1: Vec<_> = iterator.take(5).collect();
```
And at the end of the function:
```rust
        // Check that all returned keys are prefixed by requested seek key.
        let _ = iterator.count();
```

blasrodri replied:
`.take()` consumes `self`.

mina86 replied:
Collect all and truncate vector afterwards then I suppose.

blasrodri replied:
where? iterator can't be used after the `take`

blasrodri replied:
>  "‘{key:?x}’ does not start with ‘{start_key:?x}’");

this formatting isn't correct.

mina86 replied:
Yeah, I never remember.  It’s either `:?x` or `:x?`.  If the former doesn’t work than it's the latter.

pompon0 committed:
refactored PeerActor::new (#7503)

This is part of a preparation to support TIER1 peers.
I've moved as much logic as possible inside the constructor, so that the number of arguments is minimal. I've also added a PeerActor::Config which is passed to the constructor, because PeerActor::new cannot return an error (due to Actix API constraints). Also moved accepting new connections from PeerManagerActor to NetworkState.

jakmeier committed:
feat(estimator): option to select metric (#7532)

Allow to select only one metric instead of always running both
in `estimator-warehouse estimate`.

This is necessary because `icount` estimations have become very slow.

mina86 commented on core/store/src/trie/iterator.rs:
```diff
@@ -130,6 +135,9 @@ impl<'a> TrieIterator<'a> {
                         hash = *child.unwrap_hash();
                         *status = CrumbStatus::At;
                         self.key_nibbles.extend(existing_key.iter());
+                        if key.is_empty() {
+                            *prefix_boundary = true;
+                        }
```
Two things here.  Firstly, we want to set the prefix_boundary flag only if we were called from `seek_prefix`.  `seek_nibble_slice` needs `is_prefix_seek: bool` argument or alternatively we can add `seek_nibble_slice_impl` method with the implementation and have `seek_prefix` and `seek_nibble_slice` call that.

Secondly, we want to set the flag on all but the last node I think.  This way we’ll visit the node that we’ve reached but then as soon as we exit it we’ll exit all the other nodes.  This should be doable with the following code structure:

```rust
        let mut prev_prefix_boundary = &mut false;
        loop {
            *prev_prefix_boundary = is_seek_prefix;
            self.descend_into_node(&hash)?;
            let Crumb { status, node, prefix_boundary } = self.trail.last_mut().unwrap();
            prev_prefix_boundary = prefix_boundary;
            match &node.node { /* ... */ }
        }
    }
```

mina86 commented on core/store/src/trie/mod.rs:
```diff
@@ -1051,7 +1051,7 @@ mod tests {
-                iterator.seek(&query).unwrap();
+                iterator.seek_prefix(&query).unwrap();
                 if let Some(Ok((key, _))) = iterator.next() {
                     assert!(key >= query);
```
```suggestion
                    assert!(key.starts_with(query),
                            "‘{key:?x}’ does not start with ‘{query:?x}’");
```

blasrodri replied:
changing to 
```rs
assert!(key.starts_with(&query), ....
```
makes the test panic

mina86 commented on core/store/src/trie/mod.rs:
```diff
@@ -895,7 +895,7 @@ mod tests {
         let mut other_iter = trie.iter().unwrap();
-        other_iter.seek(b"r").unwrap();
+        other_iter.seek_prefix(b"r").unwrap();
         assert_eq!(other_iter.next().unwrap().unwrap().0, b"x".to_vec());
```
```suggestion
        assert_eq!(None, other_iter.next().unwrap());
```

And how about also adding a case which seeks to `b"x"` and test verifies that only `b"x"` entry is returned.

blasrodri replied:
>         assert_eq!(None, other_iter.next().unwrap());
Doesn't compile. 

> And how about also adding a case which seeks to b"x" and test verifies that only b"x" entry is returned.

I don't understand what you want to test here.

mina86 replied:
I guess `assert_eq!(None, other_iter.next());`.  Maybe it’s worth adding a helper functions here.  Something like:

```
let assert_has_next = |want| {
    assert_eq!(Some(want), other_iter.next().map(|item| item.unwrap().0).as_deref());
};
```

I’m not 100% sure if it’ll compile.

That when seeking to ‘x’ only ‘x’ is returned.  Since now seeking to ‘r’ doesn’t return anything (since ‘r’ is not a prefix of any of the keys).

blasrodri replied:
updated

matklad committed:
feat: track current sizes of network buffers (#7530)

When profiling memory usage, I've noticed non-trivial amount of space is occupied by net IO buffers. So, let's add metrics to track their size!

We use SockAddr, rather than `PeerId`, as a label here, as we don't know peer ids for outgoing connections. 

I don't worry much about perf hit here -- we more or less do a syscall here, that should dominate metric bump. 

The implementation is a bit hacky, but the "proper" fix would probably require unforking the FramedRead/Write infrastructure.

mzhangmzz committed:
Add catchup info to debug page (#7534)

Debug page now has as section for catchup information

<img width="785" alt="Screen Shot 2022-09-02 at 11 58 39 AM" src="https://user-images.githubusercontent.com/34969888/188193706-291ae020-888d-4215-ac75-bd43056d280b.png">

Catchup info will also be printed as part of stats in log. 
```
2022-09-02T15:58:30.730076Z  INFO stats: Catchups
Sync block `CZL3BCD15QpZ4FLqWu7vAJ97JsVHJ1ngtZQupfhiVaL6`@61
Shard sync status: Shard 0 done, Shard 1 split applying total parts 3 done 1, Shard 2 split applying not started, Shard 3 split applying not started
Next blocks to catch up: `CZL3BCD15QpZ4FLqWu7vAJ97JsVHJ1ngtZQupfhiVaL6`@61
```

mzhangmzz committed:
Add RejectBlocksWithOutdatedProtocolVersions to the next shardnet release (#7538)

* add feature RejectBlocksWithOutdatedProtocolVersions to the next shardnet release

* address comments

mina86 committed:
network: deprecate near_peer_message_received_total (#7548)

Deprecate near_peer_message_received_total Prometheus metric in favour
of aggregating near_peer_message_received_by_type_total metric instead.

matklad committed:
chore: remove dead code (#7539)

This ... is somewhat questionable API, let's remove it if we are not
using it.

Was added and removed in

* 061161bd7b8566bea541537d6d98177ee6246827
* 1198ec789abcdf71fecbe37b3be81ad00417146d

mina86 committed:
Prefer `&[T]` to `&Vec<T>` (#7547)

Taking slice as argument is more flexible than taking reference to
a vector since the caller does not need to have to own the data.
In addition, slices also work if caller owns a boxed slice which
would need to be converted into a vector if function required
reference to a vector.  Slices also allow to take sub-slices easily.

The only remaining place is Store::push_component in near-network crate.
This is needed because near-network’s schema’s StoreUpdate::set doesn’t
work with borrowed types.  More generic magic is needed to make it work.

matklad committed:
refactor: merge near-logging-utils into o11y (#7424)

Separate crate for test-only logging utils doesn't really pull its
weight, let's just add a `testonly` module to the general o11y crate.

Especially given that there already were some prod dependencies on this
crate.

mina86 committed:
store: allow configuring RocksDB with support for subset of columns (#7544)

With cold storage approaching, it’s gonig to be necessary to be able to
open RocksDB instance with only a subset of defined columns.  Change how
RocksDB deals with column family handles.  Specifically, get_cf_handles
now accepts an iterator over DBCol values as argument and cf_handle will
return an error if trying to get a handle for a column that wasn’t set
up.

vmarkushin commented:
Hey @mina86, could you review the changes we've just pushed? It's not 100% correct, but when trying to fix the `seek_nibble_slice` implementation we got stuck. `test_iterator_seek` is running infinitely 1/5 times (randomly) and it's not clear how to fix this because it requires a more deep understanding of how the iterator works, for which we don't have time, unfortunately, so, please help.

```diff
Index: core/store/src/trie/mod.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/core/store/src/trie/mod.rs b/core/store/src/trie/mod.rs
--- a/core/store/src/trie/mod.rs	(revision 7f44858bf6061e41822d6eb99a786843a964749b)
+++ b/core/store/src/trie/mod.rs	(date 1662386318080)
@@ -896,8 +896,8 @@
 
         let mut other_iter = trie.iter().unwrap();
         other_iter.seek_prefix(b"r").unwrap();
-        assert_eq!(other_iter.next().unwrap().unwrap().0, b"x".to_vec());
-        assert_eq!(other_iter.next().unwrap().unwrap().0, b"y".to_vec());
+        // assert_eq!(other_iter.next().unwrap().unwrap().0, b"x".to_vec());
+        // assert_eq!(other_iter.next().unwrap().unwrap().0, b"y".to_vec());
         assert_eq!(other_iter.next(), None);
         other_iter.seek_prefix(b"x").unwrap();
         assert_eq!(other_iter.next().unwrap().unwrap().0, b"x".to_vec());
@@ -1059,7 +1059,8 @@
                 let mut iterator = trie.iter().unwrap();
                 iterator.seek_prefix(&query).unwrap();
                 if let Some(Ok((key, _))) = iterator.next() {
-                    assert!(key >= query);
+                    assert!(key.starts_with(&query), "‘{key:x?}’ does not start with ‘{query:x?}’");
+                    // assert!(key.starts_with(&query), "‘{key:x?}’ does not start with ‘{query:x?}’");
                 }
             }
         }
Index: core/store/src/trie/iterator.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/core/store/src/trie/iterator.rs b/core/store/src/trie/iterator.rs
--- a/core/store/src/trie/iterator.rs	(revision 7f44858bf6061e41822d6eb99a786843a964749b)
+++ b/core/store/src/trie/iterator.rs	(date 1662386844161)
@@ -107,13 +107,22 @@
             self.descend_into_node(&hash)?;
             let Crumb { status, node, prefix_boundary } = self.trail.last_mut().unwrap();
             prev_prefix_boundary = prefix_boundary;
-            match &node.node {
+            match dbg!(&node.node) {
                 TrieNode::Empty => break,
                 TrieNode::Leaf(leaf_key, _) => {
                     let existing_key = NibbleSlice::from_encoded(leaf_key).0;
-                    if existing_key < key {
-                        self.key_nibbles.extend(existing_key.iter());
-                        *status = CrumbStatus::Exiting;
+                    if is_prefix_seek {
+                        // if existing_key == key {
+                        *prefix_boundary = true;
+                        *status = CrumbStatus::Exiting;
+                        self.key_nibbles.extend(existing_key.iter());
+                        break;
+                        // }
+                    } else {
+                        if existing_key < key {
+                            self.key_nibbles.extend(existing_key.iter());
+                            *status = CrumbStatus::Exiting;
+                        }
                     }
                     break;
                 }
@@ -128,18 +137,29 @@
                             hash = *child.unwrap_hash();
                             key = key.mid(1);
                         } else {
+                            println!("b");
+                            *prefix_boundary = true;
+                            // *prev_prefix_boundary = is_prefix_seek;
                             break;
                         }
                     }
                 }
                 TrieNode::Extension(ext_key, child) => {
                     let existing_key = NibbleSlice::from_encoded(ext_key).0;
+
                     if key.starts_with(&existing_key) {
                         key = key.mid(existing_key.len());
                         hash = *child.unwrap_hash();
                         *status = CrumbStatus::At;
                         self.key_nibbles.extend(existing_key.iter());
                     } else {
+                        if is_prefix_seek {
+                            *prefix_boundary = true;
+                            *status = CrumbStatus::Exiting;
+                            // self.key_nibbles.extend(existing_key.iter());
+                            break;
+                        }
+
                         if existing_key < key {
                             *status = CrumbStatus::Exiting;
                             self.key_nibbles.extend(existing_key.iter());

```

mina86 committed:
store: make pre-write check thresholds into constants (#7546)

check_free_space_interval and free_space_threshold fields of RocksDB
struct have always the same value.  Turn them into constants.

mina86 commented:
Hmm, then again, I suppose test_iterator_seek doesn’t build that big of a trie plus it only checks a single key.

mina86 committed:
store: extend test_iterator_seek test (#7555)

Extend test_iterator_seek by adding verification of known existing
keys.

matklad committed:
refactor: merge metrics crate into o11y (#7540)

Metrics *are* part of o11y package, it's much better organizationally to
not split this stuff into two crates.

mina86 committed:
store: move DbVersion and DB_VERSION to near-store (#7552)

The database version definitions make more sense inside of near-store
than near-primitives.  Especially since majority of usage of them
is in near-store and the definitions aren’t used for anything specific
to the chain or other operations not involving the storage.

While at it, change RocksDB::get_version such that it checks whether the
database exists (and returns None if it doesn’t) and opens the database
with only DbVersion column configured (so that’ll properly read version
from old databases without modifying them).

akhi3030 committed:
refactor: use a simple `if` instead of `match` + `if` (#7556)

Co-authored-by: Aleksey Kladov <aleksey.kladov@gmail.com>

mina86 commented:
By the way, could you make this a PR against near/nearcore repository and [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)? It’s going to be easier for me to follow up with merges and quick fixes.

blasrodri commented:
Yes. Please @seunlanlege

mina86 commented:
> test_iterator_seek is running infinitely 1/5 times (randomly)

~~I actually cannot reproduce it.  I fetched this branch, merged upstream’s master and the test works perfectly fine:~~

Nevermind.  I was fetching the wrong code.

blasrodri commented:
> By the way, could you make this a PR against near/nearcore repository and [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)? It’s going to be easier for me to follow up with merges and quick fixes.

I cannot make this PR against near/nearcore because [mpt-proof-rework](https://github.com/ComposableFi/nearcore/tree/mpt-proof-rework) isn't part of that repo. Will allow edit access from maintainers

mina86 commented:
Yes, it needs to be rebased with just the changes related to seek_prefix.  Turns out *this* is the difficult part of the proof feature so it needs to be handled as a separate change.

mina86 commented:
Turns out this is a bit more complicated than I’ve initially imagine, but this should help:

```diff
@@ -62,26 +67,50 @@ impl<'a> TrieIterator<'a> {
     }

     /// Position the iterator on the first element with key => `key`.
-    pub fn seek<K: AsRef<[u8]>>(&mut self, key: K) -> Result<(), StorageError> {
-        self.seek_nibble_slice(NibbleSlice::new(key.as_ref())).map(drop)
+    pub fn seek_prefix<K: AsRef<[u8]>>(&mut self, key: K) -> Result<(), StorageError> {
+        self.seek_nibble_slice(NibbleSlice::new(key.as_ref()), true).map(drop)
     }

     /// Returns the hash of the last node
     pub(crate) fn seek_nibble_slice(
         &mut self,
         mut key: NibbleSlice<'_>,
+        is_prefix_seek: bool,
     ) -> Result<CryptoHash, StorageError> {
         self.trail.clear();
         self.key_nibbles.clear();
+
+        // Checks if a key in an extension or leaf matches our search query.
+        //
+        // When doing prefix seek, this checks whether `key` is a prefix of
+        // `ext_key`.  When doing regular range seek, this checks whether `key`
+        // is no greater than `ext_key`.  If those conditions aren’t met, the
+        // node with `ext_key` should not match our query.
+        let check_ext_key = |key: &NibbleSlice, ext_key: &NibbleSlice| {
+            if is_prefix_seek {
+                ext_key.starts_with(key)
+            } else {
+                ext_key >= key
+            }
+        };
+
         let mut hash = self.trie.root;
+        let mut prev_prefix_boundary = &mut false;
         loop {
+            *prev_prefix_boundary = is_prefix_seek;
             self.descend_into_node(&hash)?;
-            let Crumb { status, node } = self.trail.last_mut().unwrap();
+            let Crumb { status, node, prefix_boundary } = self.trail.last_mut().unwrap();
+            prev_prefix_boundary = prefix_boundary;
             match &node.node {
                 TrieNode::Empty => break,
                 TrieNode::Leaf(leaf_key, _) => {
                     let existing_key = NibbleSlice::from_encoded(leaf_key).0;
-                    if existing_key < key {
+                    if !check_ext_key(&key, &existing_key) {
                         self.key_nibbles.extend(existing_key.iter());
                         *status = CrumbStatus::Exiting;
                     }
@@ -98,6 +127,7 @@ impl<'a> TrieIterator<'a> {
                             hash = *child.unwrap_hash();
                             key = key.mid(1);
                         } else {
+                            *prefix_boundary = is_prefix_seek;
                             break;
                         }
                     }
@@ -110,9 +139,9 @@ impl<'a> TrieIterator<'a> {
                         *status = CrumbStatus::At;
                         self.key_nibbles.extend(existing_key.iter());
                     } else {
-                        if existing_key < key {
+                        if !check_ext_key(&key, &existing_key) {
                             self.key_nibbles.extend(existing_key.iter());
                             *status = CrumbStatus::Exiting;
                         }
                         break;
                     }
```

When doing prefix seek, we need to check whether `!existing_key.starts_with(key)` rather than if `existing_key < key`.

There are still some test failures in the PR but this appears to fix test_iterator.

mina86 committed:
store: document migration snapshot options deprecation in change log (#7560)

pompon0 committed:
moved NetworkState to a separate module (#7550)

also adjusted a bit the internal API for sending routed messages.

mina86 committed:
runtime: remove serde implementation from VMContext (#7558)

The serde::Serialize and serde::Deserialize implementations for
the VMContext type are never used.  Get rid of them.  This makes
base58_format unused so get rid of it as well.

mina86 committed:
json: remove dependency on uuid crate (#7561)

Instead of introducing a new dependency, we can generate request id
using generate_random_string function.  The id doesn’t really matter
anyway (apart from JSON RPC specification requiring it to be present)
so any string really will work there.

mina86 committed:
store: move StoreOpener to its own module (#7559)

Just a tiny bit of refactoring separating configuration part of the
storage and code responsible for opening the storage into separate
modules.

mina86 committed:
store: make TrieNode’s debug formatting more concise (#7557)

The default Debug derive for types is quite verbose with a lot of
vertical spacing.  For example, TrieNode branches are printed as
a list with each empty branch as a None on its own line.  This
isn’t very ergonomic when trying to debug trie nodes and printing
a lot of the nodes to output.

Replace the default Debug implementation for TrieNode, NodeHandle and
ValueHandle by custom ones which are more concise and don’t waste
space on empty nodes.

blasrodri committed:
store: return raw bytes from Trie::retrieve_node (#7535)

This will be needed in future commit where we will need access to the
raw bytes of the node.  See #7509.

mina86 commented:
Could you update the PR to be onto master branch and remove proof-related changes from it? At this point it’s probably easiest to squash everything and clean it up:

```shell
git remote update
git checkout update-iterator-seek
git branch update-iterator-seek-backup
base=$(git merge-base @ upstream/master)
                      # ^^^^^^^^ if necessary, adjust ‘ustream’ to whatever the
                      # name of the near/nearcore remote is in the repository
git reset "$base"
git gui
```

With `git gui` (or `git add -p` if you prefer command line interface) you can pick only changes which relate to the seek_prefix feature. And once that’s done commit and force push:

```shell
git commit -m seek_prefix
git push origin +@:update-iterator-seek
```

blasrodri committed:
seek_prefix
